Enhance Upload Service to support multiple destinations

EugenM-SAP opened the pull request:
<!-- 
Thank you for supporting us with your Pull Request! 🙌 ❤️ 

Before submitting, please take the time to check the points below and provide some descriptive information.
* [ ] If this PR comes from a fork, please [Allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
* [ ] Set a meaningful title. Format: {task_name} (closes #{issue_number}). For example: Use logger (closes #41)
* [ ] [Link your Pull Request to an issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) (if applicable)
* [ ] Create Work In Progress [WIP] pull requests only if you need clarification or an explicit review before you can continue your work item.
* [ ] Make sure that your PR is not introducing _unncessary_ reformatting (e.g., introduced by on-save hooks in your IDE)

-->

## Checklist

* [x] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [x] Make sure `mvn -P integration-tests clean verify` runs for the whole project and, if you touched any code in the respective [service](services), ensure it can be run with `spring-boot:run`

## Description

Introduces enhancements for the Upload Service with regards to integration with the Swiss Federation Gateway (SGS):

*New dedicated table to store keys which are uploadable to the SGS (name: 'sgs_upload_key')
* New db user and role for the sgs_upload_key table. This user will be assigned to the Upload job instance that is running.
* New DB triggers to replicate diagnosis keys to sgs_upload_key as well as to remove keys from it when retention job is running.
* Extend current integration test to cover replication and retention of keys for both federation gateway tables.
* Add separate profiles for connection to SGS and EFGS ("connect-efgs", "connect-sgs"). The service has to retrieve data from different tables depending on the target system it has to upload it to. There were two aproaches considered for this purpose:
     * Parameterize the upload key queries with a dynamic tablename. This approach however has the downside of migrating the data access specific code to using JdbcTemplate (or other classic methods) instead of the current Spring Jdbc Repositories layer. This is because @Query does not support dynamic table parametrization (via SpEL for example).
     * Create separate profiles with separate repository implementations. This is the approach taken in this PR and although it requires duplication of @Query statements for each context, it decouples the SGS and EFGS connection completely which can help later on if their requirements around data model evolve separately.

EugenM-SAP committed:
Create the new key table for SGS uploads

EugenM-SAP committed:
Create user and permissions for new sgs upload table

EugenM-SAP committed:
Create the triggers for replication and removal of swiss uploadable keys

EugenM-SAP committed:
Fix tests for upload - missing user

EugenM-SAP committed:
Add an integration test for replication/retention of sgs upload keys

EugenM-SAP committed:
Split EFGS / SGS repository and test data generation implementations

EugenM-SAP committed:
Fix key replication integration test

EugenM-SAP committed:
Run integration tests for separate profiles

FelixRottler commented on setup/setup-roles.sql:
```diff
@@ -76,3 +76,12 @@ CREATE ROLE cwa_federation_download
    NOCREATEROLE
    NOREPLICATION
    IN ROLE cwa_user;
+  
```
I assume that this user is only for the sgs table. I think it would also make sense to have AC here. So that this swiss Role cannot have access to the other efgs table. What is your opinion on this?

EugenM-SAP committed:
Merge branch 'swiss-gateway-integration' into sgs-upload-enhancements

EugenM-SAP committed:
Rename from sgs to chgs throughout code and scripts

hilmarf committed:
Ubuntu 16.04 reaches the end of its LTS (#1234)

Ubuntu 16.04 reaches the end of its LTS window at the end of April 2021
and will no longer be supported by Canonical. As a result, the final
16.04 CircleCI machine image release by us will take place in April to
include the most recent security patches. We suggest upgrading to the
Ubuntu 20.04 image for continued releases past April. 2021.
